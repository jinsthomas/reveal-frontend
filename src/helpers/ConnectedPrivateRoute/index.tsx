import { isAuthenticated } from '@onaio/session-reducer';
import React from 'react';
import { connect } from 'react-redux';
import { Redirect, Route, RouteProps } from 'react-router-dom';
import { Store } from 'redux';
import { DISABLE_LOGIN_RPOTECTION, LOGIN_URL } from '../../constants';

/** interface for PrivateRoute props */
interface PrivateRouteProps extends RouteProps {
  authenticated: boolean /** is the current user authenticated */;
  disableLoginProtection: boolean /** should we disable login protection */;
  redirectPath: string /** redurect to this path is use if not authenticated */;
}

/** declare default props for PrivateRoute */
const defaultPrivateRouteProps: Partial<PrivateRouteProps> = {
  authenticated: false,
  disableLoginProtection: DISABLE_LOGIN_RPOTECTION,
  redirectPath: LOGIN_URL,
};

/** The PrivateRoute component
 * This component is a simple wrapper around Route and takes all its props in
 * addition to:
 *  1. {bool} authenticated
 *  2. {string} redirectPath
 *
 * If authenticated === true then render the component supplied
 * Otherwise redirect to the redirectPath
 */
const PrivateRoute = (props: PrivateRouteProps) => {
  const {
    component: Component,
    authenticated,
    disableLoginProtection,
    redirectPath,
    ...theOtherProps
  } = props;
  return (
    /* tslint:disable jsx-no-lambda */
    <Route
      {...theOtherProps}
      render={routeProps =>
        (authenticated === true || disableLoginProtection === true) && Component ? (
          <Component {...routeProps} />
        ) : (
          <Redirect to={redirectPath} />
        )
      }
    />
    /* tslint:enable jsx-no-lambda */
  );
};

PrivateRoute.defaultProps = defaultPrivateRouteProps;

export { PrivateRoute }; // export the un-connected component

/** Connect the component to the store */

/** map state to props */
const mapStateToProps = (state: Partial<Store>, ownProps: Partial<PrivateRouteProps>) => {
  const result = {
    authenticated: isAuthenticated(state),
  };
  Object.assign(result, ownProps);

  return result;
};

/** create connected component */

/** The ConnectedPrivateRoute component
 * This component is a simple wrapper around Route and takes all its props in
 * addition to:
 *  1. {bool} authenticated - this comes from the Redux store
 *  2. {string} redirectPath
 *
 * If authenticated === true then render the component supplied
 * Otherwise redirect to the redirectPath
 */
const ConnectedPrivateRoute = connect(
  mapStateToProps,
  null
)(PrivateRoute);

export default ConnectedPrivateRoute;
