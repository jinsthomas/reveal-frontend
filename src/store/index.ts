import { combine } from '@onaio/redux-reducer-registry';
import reducerRegistry from '@onaio/redux-reducer-registry';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { applyMiddleware, compose, createStore, Reducer } from 'redux';
import thunk from 'redux-thunk';
import users, { reducerName as usersReducer } from './ducks/users';

/** Add redux devtools to Window */
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

/** Declare history variable */
export const history = createBrowserHistory();

/** Declare type for reducer Registry */
export interface Registry {
  [key: string]: Reducer;
}

/** Function to create the store */
export function getStore(reducers: Registry) {
  /** Register each of the initial reducers */
  Object.keys(reducers).forEach(reducerName => {
    reducerRegistry.register(reducerName, reducers[reducerName]);
  });

  /** Combine reducers */
  const reducer = combine(reducerRegistry.getReducers());

  /** Add redux devtools to enhancers */
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  /** Create the store */
  return createStore(reducer, composeEnhancers(applyMiddleware(thunk, routerMiddleware(history))));
}

/** Router reducer */
const connectReducer = connectRouter(history);

/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
  router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};

/** Add users reducer to registry */
defaultReducers[usersReducer] = users;

/** The initial store for the reveal web app */
const store = getStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
  store.replaceReducer(combine(reducers));
});

export default store;
