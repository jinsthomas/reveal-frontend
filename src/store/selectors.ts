import { gateKeeperReducerName } from '@onaio/gatekeeper';
import { reducerName as sessionReducerName } from '@onaio/session-reducer';
import { Store } from 'redux';

/** get API Token from the Redux store
 * @param {Partial<Store>} state - the redux store
 */
export function getApiToken(state: Partial<Store>): string {
  const extraData = (state as any)[sessionReducerName].extraData;
  return extraData.api_token || '';
}

/** get Access Token from the Redux store
 * @param {Partial<Store>} state - the redux store
 */
export function getAccessToken(state: Partial<Store>): string {
  return (state as any)[gateKeeperReducerName].accessToken;
}
