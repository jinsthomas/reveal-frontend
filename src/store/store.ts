import { combine } from '@onaio/redux-reducer-registry';
import reducerRegistry from '@onaio/redux-reducer-registry';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { applyMiddleware, createStore, Reducer } from 'redux';
import thunk from 'redux-thunk';
import users, { reducerName as usersReducer } from './ducks/users';

export const history = createBrowserHistory();

/** Declare type for reducer Registry */
export interface Registry {
  [key: string]: Reducer;
}

/** Function to create the store */
export function getStore(reducers: Registry) {
  /** Register each of the initial reducers */
  Object.keys(reducers).forEach(reducerName => {
    reducerRegistry.register(reducerName, reducers[reducerName]);
  });
  /** Combine reducers */
  const reducer = combine(reducerRegistry.getReducers());
  /** Create the store */
  return createStore(
    reducer,
    (window as any).__REDUX_DEVTOOLS_EXTENSION__ && (window as any).__REDUX_DEVTOOLS_EXTENSION__(),
    applyMiddleware(thunk, routerMiddleware(history))
  );
}

/** Router reducer */
const connectReducer = connectRouter(history);
/** Initial reducers in the reducer registry */
const defaultReducers: Registry = {
  router: connectReducer as any /** Dirty hack because connectRouter LocationChangeAction does not extend Redux.AnyAction */,
};
// /** Add users reducer to registry */
defaultReducers[usersReducer] = users;

/** The initial store for the reveal web app */
const store = getStore(defaultReducers);

/** Set listener to add reducers to store when registered */
reducerRegistry.setChangeListener(reducers => {
  store.replaceReducer(combine(reducers));
});

export default store;
