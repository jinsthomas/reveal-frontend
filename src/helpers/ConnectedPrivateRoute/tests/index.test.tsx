import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import ConnectedPrivateRoute, { PrivateRoute } from '../';
import store from '../../../store';

const history = createBrowserHistory();

/** the test component */
const TestComponent = (props: { [key: string]: any }) => {
  return <span>{props.someProp}</span>;
};

describe('ConnectedPrivateRoute', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing when authenticated', () => {
    const props = {
      authenticated: true,
      component: TestComponent,
      disableLoginProtection: false,
      redirectPath: '/login',
      someProp: 'I love Oov' /** we want to test that this prop is passed onwards */,
      someProp2: 'hidden' /** we want to test that this prop is passed onwards */,
    };
    // Date.now = jest.fn(() => 1482363367071);
    shallow(
      <MemoryRouter>
        <PrivateRoute {...props} />
      </MemoryRouter>
    );

    const wrapper = mount(
      <MemoryRouter>
        <PrivateRoute {...props} />
      </MemoryRouter>
    );
    expect(wrapper.find('TestComponent').length).toEqual(1);
    /** We are matching only the span so that we dont deal with random generated
     * keys in the spanshot test
     */
    expect(toJson(wrapper.find('TestComponent span'))).toMatchSnapshot();
    expect(wrapper.find('TestComponent').props() as { [key: string]: any }).toMatchSnapshot({
      history: expect.any(Object),
      location: expect.any(Object),
      match: expect.any(Object),
    });
    wrapper.unmount();
  });

  it('renders without crashing when disableLoginProtection is true', () => {
    const props = {
      authenticated: true,
      component: TestComponent,
      disableLoginProtection: false,
      redirectPath: '/login',
      someProp: 'treasure island' /** we want to test that this prop is passed onwards */,
      someProp2: 'treasure' /** we want to test that this prop is passed onwards */,
    };
    // Date.now = jest.fn(() => 1482363367071);
    shallow(
      <MemoryRouter>
        <PrivateRoute {...props} />
      </MemoryRouter>
    );

    const wrapper = mount(
      <MemoryRouter>
        <PrivateRoute {...props} />
      </MemoryRouter>
    );
    expect(wrapper.find('TestComponent').length).toEqual(1);
    /** We are matching only the span so that we dont deal with random generated
     * keys in the spanshot test
     */
    expect(toJson(wrapper.find('TestComponent span'))).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders without crashing when not authenticated', () => {
    const props = {
      authenticated: false,
      component: TestComponent,
      disableLoginProtection: false,
      redirectPath: '/login',
    };
    shallow(
      <MemoryRouter>
        <PrivateRoute {...props} />
      </MemoryRouter>
    );

    const wrapper = mount(
      <MemoryRouter>
        <PrivateRoute {...props} />
      </MemoryRouter>
    );

    expect(wrapper.find('TestComponent').length).toEqual(0);
    expect(wrapper.find('PrivateRoute').length).toEqual(1);
    expect(toJson(wrapper.find('PrivateRoute'))).toMatchSnapshot();
    expect(wrapper.find('PrivateRoute').props()).toEqual(props);
    expect(toJson(wrapper.find('Redirect'))).toMatchSnapshot();
    wrapper.unmount();
  });

  it('works when connected to the redux store', () => {
    shallow(
      <Provider store={store}>
        <MemoryRouter>
          <ConnectedPrivateRoute />
        </MemoryRouter>
      </Provider>
    );

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <ConnectedPrivateRoute />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('TestComponent').length).toEqual(0);
    expect(wrapper.find('PrivateRoute').length).toEqual(1);
    expect(toJson(wrapper.find('PrivateRoute'))).toMatchSnapshot();
    expect(wrapper.find('PrivateRoute').props() as any).toMatchSnapshot({
      dispatch: expect.any(Function),
    });
    expect(toJson(wrapper.find('Redirect'))).toMatchSnapshot();
    wrapper.unmount();
  });
});
