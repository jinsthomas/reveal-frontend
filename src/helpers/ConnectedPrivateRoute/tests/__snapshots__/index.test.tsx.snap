// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectedPrivateRoute renders without crashing when authenticated 1`] = `
<span>
  I love Oov
</span>
`;

exports[`ConnectedPrivateRoute renders without crashing when authenticated 2`] = `
Object {
  "history": Any<Object>,
  "location": Any<Object>,
  "match": Any<Object>,
  "someProp": "I love Oov",
  "someProp2": "hidden",
  "staticContext": undefined,
}
`;

exports[`ConnectedPrivateRoute renders without crashing when disableLoginProtection is true 1`] = `
<span>
  treasure island
</span>
`;

exports[`ConnectedPrivateRoute renders without crashing when not authenticated 1`] = `
<PrivateRoute
  authenticated={false}
  component={[Function]}
  disableLoginProtection={false}
  redirectPath="/login"
>
  <Route
    render={[Function]}
  >
    <Redirect
      push={false}
      to="/login"
    />
  </Route>
</PrivateRoute>
`;

exports[`ConnectedPrivateRoute renders without crashing when not authenticated 2`] = `
<Redirect
  push={false}
  to="/login"
/>
`;

exports[`ConnectedPrivateRoute works when connected to the redux store 1`] = `
<PrivateRoute
  authenticated={false}
  disableLoginProtection={true}
  dispatch={[Function]}
  redirectPath="/login"
>
  <Route
    dispatch={[Function]}
    render={[Function]}
  >
    <Redirect
      push={false}
      to="/login"
    />
  </Route>
</PrivateRoute>
`;

exports[`ConnectedPrivateRoute works when connected to the redux store 2`] = `
Object {
  "authenticated": false,
  "disableLoginProtection": true,
  "dispatch": Any<Function>,
  "redirectPath": "/login",
}
`;

exports[`ConnectedPrivateRoute works when connected to the redux store 3`] = `
<Redirect
  push={false}
  to="/login"
/>
`;
